#### Exercici 1. Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

NOTA: Partim del conjunt Train i Test analitzats, preprocessats i transformats en el Sprint 5.

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
#Conjunt Train analitzat, preprocessat i transformat en el Sprint 5
housing_ML_train = pd.read_excel('/Users/juliacorominasfont/Desktop/DATA ANALYST/Bootcamp IT Academy/Tasques Sprints/Sprint_6_ML_Regressions/housing_train_transformed.xlsx')
housing_ML_train.head(5)

Unnamed: 0      CRIM    ZN        RM   PTRATIO          B     LSTAT  \
0         477  4.729102  0.00 -1.214070  0.441176  -2.062123  1.456771   
1          15  0.117250  0.00 -0.503853  0.676471   0.212745 -0.255729   
2         332 -0.072634  1.75 -0.239866 -0.529412  -1.432516 -0.322396   
3         423  2.174908  0.00 -0.143384  0.441176 -19.168495  1.288021   
4          19  0.148777  0.00 -0.647236  0.676471  -0.017503  0.036979   

      INDUS       NOX       AGE  ...  CHAS  RAD_2  RAD_24  RAD_3  RAD_4  \
0  0.642963  0.471193  0.972194  ...     0      0       1      0      0   
1  0.274074  0.314815  0.552008  ...     0      0       0      0      1   
2  0.197037  0.108848  0.210093  ...     0      0       0      0      0   
3  0.642963  0.471193  0.846550  ...     0      0       1      0      0   
4  0.274074  0.314815  0.685891  ...     0      0       0      0      1   

   RAD_5  RAD_6  RAD_7  RAD_8  MEDV  
0      0      0      0      0  12.0  
1      0      0      0      0  19.9  
2      0      0      0      0  19.4  
3      0      0      0      0  13.4  
4      0      0      0      0  18.2  

[5 rows x 21 columns]

In [4]:
#Eliminació de columnes no rellevants
housing_ML_train.drop('Unnamed: 0', axis=1, inplace=True)
housing_ML_train.head(5)

CRIM    ZN        RM   PTRATIO          B     LSTAT     INDUS  \
0  4.729102  0.00 -1.214070  0.441176  -2.062123  1.456771  0.642963   
1  0.117250  0.00 -0.503853  0.676471   0.212745 -0.255729  0.274074   
2 -0.072634  1.75 -0.239866 -0.529412  -1.432516 -0.322396  0.197037   
3  2.174908  0.00 -0.143384  0.441176 -19.168495  1.288021  0.642963   
4  0.148777  0.00 -0.647236  0.676471  -0.017503  0.036979  0.274074   

        NOX       AGE       TAX  CHAS  RAD_2  RAD_24  RAD_3  RAD_4  RAD_5  \
0  0.471193  0.972194  0.914122     0      0       1      0      0      0   
1  0.314815  0.552008  0.229008     0      0       0      0      1      0   
2  0.108848  0.210093  0.223282     0      0       0      0      0      0   
3  0.471193  0.846550  0.914122     0      0       1      0      0      0   
4  0.314815  0.685891  0.229008     0      0       0      0      1      0   

   RAD_6  RAD_7  RAD_8  MEDV  
0      0      0      0  12.0  
1      0      0      0  19.9  
2      0      0      0  19.4  
3      0      0      0  13.4  
4      0      0      0  18.2

In [5]:
#Variables X i y
X_train = housing_ML_train.drop('MEDV', axis=1)
y_train = housing_ML_train['MEDV']

In [6]:
#Conjunt Test analitzat, preprocessat i transformat en el Sprint 5
housing_ML_test = pd.read_excel('/Users/juliacorominasfont/Desktop/DATA ANALYST/Bootcamp IT Academy/Tasques Sprints/Sprint_6_ML_Regressions/housing_test_transformed.xlsx')
housing_ML_test.head(5)

Unnamed: 0      CRIM   ZN        RM   PTRATIO         B     LSTAT  \
0         173 -0.054336  0.0  0.276047 -0.617647  0.206829 -0.196354   
1         274 -0.065657  2.0  0.734338 -0.323529  0.275854 -0.770313   
2         491 -0.049863  0.0 -0.304188  0.411765 -0.058918  0.744271   
3          72 -0.054380  0.0 -0.194305  0.147059 -0.019475 -0.563021   
4         452  1.546930  0.0  0.116583  0.441176 -0.306422  0.660937   

      INDUS       NOX       AGE  ...  CHAS  RAD_2  RAD_24  RAD_3  RAD_4  \
0  0.122593  0.257202  0.836251  ...     0      0       0      0      0   
1  0.210000  0.127572  0.308960  ...     1      0       0      0      1   
2  1.000000  0.460905  0.987642  ...     0      0       0      0      1   
3  0.372963  0.057613  0.050463  ...     0      0       0      0      1   
4  0.642963  0.674897  0.915551  ...     0      0       1      0      0   

   RAD_5  RAD_6  RAD_7  RAD_8  MEDV  
0      1      0      0      0  23.6  
1      0      0      0      0  32.4  
2      0      0      0      0  13.6  
3      0      0      0      0  22.8  
4      0      0      0      0  16.1  

[5 rows x 21 columns]

In [7]:
#Eliminació de columnes no rellevants
housing_ML_test.drop('Unnamed: 0', axis=1, inplace=True)
housing_ML_test.head(5)

CRIM   ZN        RM   PTRATIO         B     LSTAT     INDUS       NOX  \
0 -0.054336  0.0  0.276047 -0.617647  0.206829 -0.196354  0.122593  0.257202   
1 -0.065657  2.0  0.734338 -0.323529  0.275854 -0.770313  0.210000  0.127572   
2 -0.049863  0.0 -0.304188  0.411765 -0.058918  0.744271  1.000000  0.460905   
3 -0.054380  0.0 -0.194305  0.147059 -0.019475 -0.563021  0.372963  0.057613   
4  1.546930  0.0  0.116583  0.441176 -0.306422  0.660937  0.642963  0.674897   

        AGE       TAX  CHAS  RAD_2  RAD_24  RAD_3  RAD_4  RAD_5  RAD_6  RAD_7  \
0  0.836251  0.208015     0      0       0      0      0      1      0      0   
1  0.308960  0.127863     1      0       0      0      1      0      0      0   
2  0.987642  1.000000     0      0       0      0      1      0      0      0   
3  0.050463  0.225191     0      0       0      0      1      0      0      0   
4  0.915551  0.914122     0      0       1      0      0      0      0      0   

   RAD_8  MEDV  
0      0  23.6  
1      0  32.4  
2      0  13.6  
3      0  22.8  
4      0  16.1

In [8]:
#Variables X i y
X_test = housing_ML_test.drop('MEDV', axis=1)
y_test = housing_ML_test['MEDV']

### Models

#### Decision Tree Regression

In [9]:
from sklearn.tree import DecisionTreeRegressor

dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

In [10]:
predictions_dt = dt.predict(X_test)

#### Random forest

In [11]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [12]:
predictions_rf = rf.predict(X_test)

#### Exercici 2. Compara’ls en base al MSE (Mean Square Error) i al R2.

- MSE (error quadràtic mitjà): És una funció de risc que mesura lo aprop que està una línia de regressió d'un conjunt de punts de dades. Ajuda a quantificar la precisió del model. És convenient utilitzar aquesta mesura si la variable objectiu està normalment distribuïda i es vol que els errors grans es penalitzin més que els petits.

- R2 (coeficient de determinació): La puntuació R2 indica si el model és capaç d'explicar bé la variabilitat de la variable objectiu. Cal tenir en compte que si el R2 és molt alt pot ser indicatiu d'un sobreajustament, si no es compleixen altres assumpcions del model.

In [32]:
from sklearn.metrics import mean_squared_error
print('MSE del Decision Tree Regressor:', mean_squared_error(y_test, predictions_dt))
print('MSE del Random Forest:', mean_squared_error(y_test, predictions_rf))

MSE del Decision Tree Regressor: 20.870294117647052
MSE del Random Forest: 10.2297485490196


El model Random Forest té un MSE més petit que el Decision Tree Regression. Això ens indica que l'error del Random Forest és menor i, per tant, que té un millor rendiment predictiu. Dit d'una altra manera, el model Random Forest és més acurat a l'hora de fer les prediccions.

Cal tenir en compte que el MSE es mesura en les unitats quadrades de la variable objectiu, és a dir, per obtenir una mesura en les mateixes unitats que la variable objectiu es podria utilitzar l'arrel quadrada de l'MSE.

In [33]:
from sklearn.metrics import r2_score
print('R2 del Decision Tree Regressor:', r2_score(y_test, predictions_dt))
print('R2 del Random Forest:', r2_score(y_test, predictions_rf))

R2 del Decision Tree Regressor: 0.7154068259096871
R2 del Random Forest: 0.8605042845443384


Els dos models tenen un R2 bastant pròxim a 1, el que vol dir que els dos semblen funcionar bé a l'hora d'explicar la variabilitat de la variable objectiu (MEDV). Els resultats indiquen que el model Decision Tree Regression explica el voltant del 71,54% de la variabilitat de la variable target, mentre que el model Random Forest és capaç d'explicar el 86,05%. Per tant, el segon és més capaç d'identificar patrons en el conjunt de dades.

#### Exercici 3. Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

In [34]:
from sklearn.model_selection import GridSearchCV

# Definició de la cuadrícula de paràmetres pel DecisionTreeRegressor
param_grid_dt = {
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4, 8],
    'random_state': [42]
}

# Definició de la cuadrícula de paràmetres pel RandomForestRegressor
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4, 8],
    'random_state': [42]
}

dt_regressor = DecisionTreeRegressor()
rf_regressor = RandomForestRegressor()

# Ús del GridSearchCV per trobar els millors paràmetres pel DecisionTreeRegressor
grid_search_dt = GridSearchCV(dt_regressor, param_grid_dt, cv=5, scoring='neg_mean_squared_error')
grid_search_dt.fit(X_train, y_train)

# Ús del GridSearchCV per trobar els millors paràmetres pel RandomForestRegressor
grid_search_rf = GridSearchCV(rf_regressor, param_grid_rf, cv=5, scoring='neg_mean_squared_error')
grid_search_rf.fit(X_train, y_train)

print("Millors paràmetres pel  Decision Tree Regressor:", grid_search_dt.best_params_)
print("Millors paràmetres pel Random Forest:", grid_search_rf.best_params_)

Millors paràmetres pel  Decision Tree Regressor: {'max_depth': 5, 'min_samples_leaf': 8, 'min_samples_split': 2, 'random_state': 42}
Millors paràmetres pel Random Forest: {'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100, 'random_state': 42}


In [35]:
# Prediccions amb els nous models
predictions_best_dt = grid_search_dt.predict(X_test)
predictions_best_rf = grid_search_rf.predict(X_test)

# Evaluació dels models
print("MSE del nou  Decision Tree Regressor:", mean_squared_error(y_test, predictions_best_dt))
print('R2 del nou Decision Tree Regressor:', r2_score(y_test, predictions_best_dt))
print("\nMSE del nou Random Forest:", mean_squared_error(y_test, predictions_best_rf))
print('R2 del nou Random Forest:', r2_score(y_test, predictions_best_rf))

MSE del nou  Decision Tree Regressor: 16.594428070848036
R2 del nou Decision Tree Regressor: 0.773713732529397

MSE del nou Random Forest: 11.738112006565098
R2 del nou Random Forest: 0.8399358181085089


Veiem que el rendiment del Decision Tree Regressor ha millorat, ha disminuït el MSE i ha augmentat el R2. En el case del Random Forest, el rendiment ha empitjorat, doncs ha augmentat el MSE i ha disminuït el r2.

#### Exercici 4. Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).

In [36]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import make_scorer

#Variables X i y del conjunt (ja que fins ara teniem els Train i Test sets)
X = pd.concat([X_train, X_test])
y = pd.concat([y_train, y_test])

scorer_r2 = make_scorer(r2_score)

# DecisionTreeRegressor
dt_cv = DecisionTreeRegressor(random_state=42)
dt_scores = cross_val_score(dt_cv, X, y, cv=5, scoring='neg_mean_squared_error')
dt_mse_scores = -dt_scores
dt_r2_scores = cross_val_score(dt_cv, X, y, cv=5, scoring=scorer_r2)

# Random Forest Regressor
rf_cv = RandomForestRegressor(random_state=42)
rf_scores = cross_val_score(rf_cv, X, y, cv=5, scoring='neg_mean_squared_error')
rf_mse_scores = -rf_scores 
rf_r2_scores = cross_val_score(rf_cv, X, y, cv=5, scoring=scorer_r2)

print("MSE del Decision Tree Regressor amb CV:", np.mean(dt_mse_scores))
print("R2 del Decision Tree Regressor amb CV:", np.mean(dt_r2_scores))
print("\nMSE del Random Forest amb CV:", np.mean(rf_mse_scores))
print("R2 del Random Forest amb CV:", np.mean(rf_r2_scores))

MSE del Decision Tree Regressor amb CV: 22.20947835371772
R2 del Decision Tree Regressor amb CV: 0.7369583190548619

MSE del Random Forest amb CV: 12.637621859813626
R2 del Random Forest amb CV: 0.8496496800249533


Si comparem els resultats amb l'exercici anterior, veiem que en ambdós casos el MSE ha augmentat, el que vol dir que ha augmentat l'error del model i, per tant, ha disminuït el seu rendiment. 
En canvi, pel que fa al R2, veiem que en el cas del Decision Treee ha disminuït del 77,37% al 73,7%, i en el cas del Random Forest ha augmentat del 84% al 84,97%. Tot i així, els canvis són poc significatius. 

#### Exercici 5. No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

In [25]:
X_no_rm = X.drop(['RM'], axis=1) 

In [26]:
from sklearn.model_selection import train_test_split
X_train_no_rm, X_test_no_rm, y_train_no_rm, y_test_no_rm = train_test_split(X_no_rm, y, test_size=0.2, random_state=42)

print('Forma del Training set: X={}, y={}'.format(X_train_no_rm.shape, y_train_no_rm.shape))
print('Forma del Test set: X={}, y={}'.format(X_test_no_rm.shape, y_test_no_rm.shape))

Forma del Training set: X=(404, 18), y=(404,)
Forma del Test set: X=(102, 18), y=(102,)


In [27]:
#Decision Tree Regressor
dt_no_rm = DecisionTreeRegressor(random_state=42)
dt_no_rm.fit(X_train_no_rm, y_train_no_rm)

DecisionTreeRegressor(random_state=42)

In [28]:
predictions_dt_no_rm = dt_no_rm.predict(X_test_no_rm)

In [29]:
#Random Forest
rf_no_rm = RandomForestRegressor(random_state=42)
rf_no_rm.fit(X_train_no_rm, y_train_no_rm)

RandomForestRegressor(random_state=42)

In [30]:
predictions_rf_no_rm = rf_no_rm.predict(X_test_no_rm)

In [37]:
#Avaluació dels models
print('MSE del Decision Tree Regressor sense RM:', mean_squared_error(y_test_no_rm, predictions_dt_no_rm))
print('R2 del Decision Tree Regressor sense RM:', r2_score(y_test_no_rm, predictions_dt_no_rm))
print('\nMSE del Random Forest sense RM:', mean_squared_error(y_test_no_rm, predictions_rf_no_rm))
print('R2 del Random Forest sense RM:', r2_score(y_test_no_rm, predictions_rf_no_rm))

MSE del Decision Tree Regressor sense RM: 37.923725490196084
R2 del Decision Tree Regressor sense RM: 0.6355729026101308

MSE del Random Forest sense RM: 30.054735480392168
R2 del Random Forest sense RM: 0.711189766501946


Si comparem els resultats amb els de l'exercici 3, veiem que el rendiment dels models ha empitjorat considerablement. D'una banda, el MSE ha augmentat en els dos casos, de 16.59 a 37.92 en el Decision Tree i de 11.73 a 30.05 en el Random Forest. I, de l'altra, el R2 ha disminuït; de 0.77 a 0.64 pel primer i de 0.84 a 0.71 pel segon.

Per tant, veiem que quan eliminem la variable del nombre d'habitacions, RM, obtenim un major error i, per tant, un pitjor rendiment.